_posts/2019-11-19-demo-fraud-detection.md

```diff
@@ -0,0 +1,223 @@
+
+```java
+DataStream<Alert> alerts =
+    transactions
+        .process(new DynamicKeyFunction())
+        .keyBy((keyed) -> keyed.getKey());
+        .process(new DynamicAlertFunction())
+
+        //.process(/* actual calculations and alerting */)
+```
+
+By applying `DynamicKeyFunction` we are effectively copying events for performing parallel per-rule evaluation within a Flink cluster. By doing so, we achieve an important property - horizontal scalability of rules' processing. Our system will be capable of handling more rules by adding more servers to the cluster, i.e. increasing the parallelism. This property is achieved at the cost of data duplication, which might become an issue depending on the specific set of parameters, such as incoming data rate, available network bandwidth, event payload size etc. In a real-life scenario, additional optimizations can be applied, such as combined evaluation of rules which have the same `groupingKeyNames`, or a filtering layer, which would strip events of all the fields that are not required for processing of a particular rule.
+
+### Summary:
+
+In this blog post we have discussed the motivation behind supporting dynamic, runtime changes to a Flink application by looking at a sample use case - a Fraud Detection engine. We have described the overall architecture and interactions between its components as well as provided references for building and running a demo Fraud Detection application in a dockerized setup. We have proceeded with diving into the details of implementing a  **dynamic data partitioning pattern** as the first underlying building block which enables functionality of a flexible fraud detection engine.
+
+For keeping focus on describing the core mechanics of the pattern, we have kept the complexity of the DSL and the underlying rules engine to a minimum. It is, however, easy to think about extensions allowing more sophisticated rules definitions, including filtering of certain events, logical rules chaining and other more advanced functionality.
+
+In the second part of this series, we will describe how the rules actually make their way into the running Fraud Detection engine. Additionally, we will go over the implementation details of the main processing function of the pipeline - _DynamicAlertFunction()_.
+
+<center>
+<img src="{{ site.baseurl }}/img/blog/2019-11-19-demo-fraud-detection/end-to-end.png" width="800px" alt="Figure 4: End-to-end pipeline"/>
+<br/>
+<i><small>Figure 4: End-to-end pipeline</small></i>
+</center>
+<br/>
+
+In the next article we will see how Flink's broadcast streams can be utilized to help steer the processing within the Fraud Detection engine at runtime (Dynamic Application Updates pattern).
```

Comment:
```suggestion
In the next article, we will see how Flink's broadcast streams can be utilized to help steer the processing within the Fraud Detection engine at runtime (Dynamic Application Updates pattern).
```

Reply:
Some weird set of rules for this one: https://english.stackexchange.com/questions/294864/comma-after-in-section-2-in-what-follows-or-in-1999